evm/p2p/kademlia.py

```diff
@@ -0,0 +1,446 @@
+"""
+Node discovery and network formation are implemented via a kademlia-like protocol.
+The major differences are that packets are signed, node ids are the public keys, and
+DHT-related features are excluded. The FIND_VALUE and STORE packets are not implemented.
+The parameters necessary to implement the protocol are a
+bucket size of 16 (denoted k in Kademlia),
+concurrency of 3 (denoted alpha in Kademlia),
+and 8 bits per hop (denoted b in Kademlia) for routing.
+The eviction check interval is 75 milliseconds,
+request timeouts are 300ms, and
+the idle bucket-refresh interval is 3600 seconds.
+
+Aside from the previously described exclusions, node discovery closely follows system
+and protocol described by Maymounkov and Mazieres.
+"""
+import operator
+import random
+import time
+from functools import total_ordering
+
+from rlp.utils import is_integer, str_to_bytes
+
+from evm.utils.keccak import keccak
+from evm.utils.numeric import big_endian_to_int
+
+# TODO: Setup a logger using the standard logging module.
+from structlog import get_logger
+log = get_logger()
+
+k_b = 8  # 8 bits per hop
+
+k_bucket_size = 16
+k_request_timeout = 3 * 300 / 1000.      # timeout of message round trips
+k_idle_bucket_refresh_interval = 3600    # ping all nodes in bucket if bucket was idle
+k_find_concurrency = 3                   # parallel find node lookups
+k_pubkey_size = 512
+k_id_size = 256
+k_max_node_id = 2 ** k_id_size - 1
+
+
+def random_nodeid():
+    return random.randint(0, k_max_node_id)
+
+
+@total_ordering
+class Node(object):
+
+    def __init__(self, pubkey):
+        self.pubkey = pubkey
+        if k_id_size == 512:
+            self.id = big_endian_to_int(pubkey)
+        else:
+            assert k_id_size == 256
```

Comment:
Assert statements can easily be disabled: https://stackoverflow.com/questions/1273211/disable-assertions-in-python

I've preferred explicit raising of a `ValueError` or `TypeError` instead.

Reply:
These asserts are there to help debugging while I'm working on this code; there's a TODO at the top of the file to remind me to go through all the asserts and either replace them with exceptions or drop them ;-)